test/e2e_node/image_gc_test.go

```diff
@@ -0,0 +1,107 @@
+	internalapi "k8s.io/cri-api/pkg/apis"
+	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1"
+	kubefeatures "k8s.io/kubernetes/pkg/features"
+	kubeletconfig "k8s.io/kubernetes/pkg/kubelet/apis/config"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+	admissionapi "k8s.io/pod-security-admission/api"
+
+	"github.com/onsi/ginkgo/v2"
+	"github.com/onsi/gomega"
+)
+
+var _ = SIGDescribe("ImageGarbageCollect", framework.WithSerial(), framework.WithNodeFeature("GarbageCollect"), func() {
+	f := framework.NewDefaultFramework("image-garbage-collect-test")
+	f.NamespacePodSecurityLevel = admissionapi.LevelPrivileged
+	var is internalapi.ImageManagerService
+	ginkgo.BeforeEach(func() {
+		var err error
+		_, is, err = getCRIClient()
+		framework.ExpectNoError(err)
+	})
+	ginkgo.AfterEach(func() {
+		framework.ExpectNoError(PrePullAllImages())
+	})
+	ginkgo.Context("when ImageMaximumGCAge is set", func() {
+		tempSetCurrentKubeletConfig(f, func(ctx context.Context, initialConfig *kubeletconfig.KubeletConfiguration) {
+			initialConfig.ImageMaximumGCAge = metav1.Duration{Duration: time.Duration(time.Minute * 1)}
+			initialConfig.ImageMinimumGCAge = metav1.Duration{Duration: time.Duration(time.Second * 1)}
+			if initialConfig.FeatureGates == nil {
+				initialConfig.FeatureGates = make(map[string]bool)
+			}
+			initialConfig.FeatureGates[string(kubefeatures.ImageMaximumGCAge)] = true
+		})
+		ginkgo.It("should GC unused images", func(ctx context.Context) {
+			pod := innocentPod()
+			e2epod.NewPodClient(f).CreateBatch(ctx, []*v1.Pod{pod})
+
+			_, err := is.PullImage(context.Background(), &runtimeapi.ImageSpec{Image: agnhostImage}, nil, nil)
+			framework.ExpectNoError(err)
+
+			allImages, err := is.ListImages(context.Background(), &runtimeapi.ImageFilter{})
+			framework.ExpectNoError(err)
+
+			e2epod.NewPodClient(f).DeleteSync(ctx, pod.ObjectMeta.Name, metav1.DeleteOptions{}, e2epod.DefaultPodDeletionTimeout)
+
+			// Even though the image gc max timing is less, we are bound by the kubelet's
+			// ImageGCPeriod, which is hardcoded to 5 minutes.
+			gomega.Eventually(ctx, func() int {
+				gcdImageList, err := is.ListImages(context.Background(), &runtimeapi.ImageFilter{})
+				framework.ExpectNoError(err)
+				return len(gcdImageList)
+			}, 6*time.Minute, 30*time.Second).Should(gomega.BeNumerically("<", len(allImages)))
+		})
+		ginkgo.It("should not GC unused images prematurely", func(ctx context.Context) {
+			pod := innocentPod()
+			e2epod.NewPodClient(f).CreateBatch(ctx, []*v1.Pod{pod})
+
+			_, err := is.PullImage(context.Background(), &runtimeapi.ImageSpec{Image: agnhostImage}, nil, nil)
+			framework.ExpectNoError(err)
+
+			allImages, err := is.ListImages(context.Background(), &runtimeapi.ImageFilter{})
+			framework.ExpectNoError(err)
+
+			e2epod.NewPodClient(f).DeleteSync(ctx, pod.ObjectMeta.Name, metav1.DeleteOptions{}, e2epod.DefaultPodDeletionTimeout)
+
+			restartKubelet(true)
```

Comment:
Why the restart?

Reply:
nvm. I see the comment on 99.

Reply:
Looking at restartKubelet, I think we are going to hit some flakes with this.

restartKubelet doesn't wait for kubelet to be ready so I think we may hit issues.

Reply:
@haircommander Maybe we should add a health check here?

Reply:
from my understanding kubelet will signal to systemd when it's ready, and after that point it has already started up, which is what systemd waits for to exit the `systemctl restart` code. Thus, I'm pretty sure the kubelet is ready after `systemctl restart`. Since other uses of this function also don't add a healthcheck, it's my preference not to. WDYT?

Reply:
My main issue is that UpdateKubeletConfig waits for the health check: https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/util.go#L216

Reply:
Maybe if it flakes we can leave this as a possible follow up.

Reply:
no need for a follow up, I've added it :)